tests/MongoDB.Bson.Tests/Serialization/SerializationTests.cs

```diff
@@ -14,30 +14,44 @@
-        [InlineData(typeof(DerivedWithoutSetterAndWithBsonElementAndWithPrivateConstructor_BaseWithoutSetter))]
-        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetterAndWithProtectedConstructor))]
-        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetter))]
-        [InlineData(typeof(DerivedWithPrivateSetterAndWithBsonElement_BaseWithoutSetter))]
-        public void Serialization_should_work_as_expected(Type testCaseType)
+        [InlineData(typeof(DerivedWithoutSetter_BaseWithoutSetter), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedWithoutSetter_BaseWithPrivateSetterAndWithProtectedConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedWithoutSetterAndWithBsonElement_BaseWithoutSetterAndWithProtectedConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"y\" : 2 }")]
+        [InlineData(typeof(DerivedWithoutSetterAndWithBsonElementAndWithPrivateConstructor_BaseWithoutSetter), new[] { 1, 2 }, "{ \"X\" : 1, \"y\" : 2 }")]
+        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetterAndWithProtectedConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetter), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedWithPrivateSetterAndWithBsonElement_BaseWithoutSetter), new[] { 1, 2 }, "{ \"X\" : 1, \"y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorAndWithBsonConstructorAttribute_AbstractBaseImmutable), new[] { 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2  }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutable), new[] { 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithAbstractProperty), new[] { 2 }, "{ \"Y\" : 2 }")]
+        [InlineData(typeof(DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithConstructor), new[] { 1 }, "{ \"X\" : 1 }")]
+        [InlineData(typeof(ClassWithTwoConstructorsWhereTheFirstIfNotFullyMatchedAndTheSecondIsFullyMatched), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2 }")]
+        [InlineData(typeof(ClassWithOneConstructorThatIsNotFullyMatched), new[] { 1, 2 }, "{ \"X\" : 1, \"Y\" : 2, \"Z\" : null }")]
```

Comment:
It's worth asking whether `Z` should be serialized or not.

We might want to make a small change to `ImmutableTypeClassMapConvention` (line 88) so that only properties that can be matched to a constructor argument (any constructor) will be mapped.

This used to not be an issue because we used to only consider the one constructor (if any) that had the same number of arguments as the class had properties.